### PR TITLE
Separate the call of different plugins in ecscni

### LINKED ISSUE #1151: ECS Agent setting up the container namespace will fail with the latest cni plugins
### Summary

Task networking will no longer work with the new changes to cni plugin aws/amazon-ecs-cni-plugins#59, as agent invoke all the plugin(`ecs-bridge, eni, ipam`) with the same runtime configuration [here](https://github.com/aws/amazon-ecs-agent/blob/1ca656c8d53785579641376f32e5167e018952b2/agent/ecscni/plugin.go#L65), where the interface name is the same for all the plugins, this cause the conflict of veth pair(created by `ecs-bridge` plugin) and eni (name was reset by `eni` plugin).

### Description

Task launched with `awsvpc` mode will fail with the latest change in aws/amazon-ecs-cni-repo.

### Expected Behavior
Setting up the container namespace correctly without error

### Observed Behavior
Task failed in `awsvpc` mode

### Supporting Log Snippets

Setting up the pause container namespace will fail with the error: 
```
Unable to configure pause container namespace, err: bridge create veth pair: unable to setup veth pair for interface: ecs-eth0: container veth name provided (ecs-eth0) already exists, task: arn:aws:ecs:us-west-2:xxxxxx:task/xxxxx
```

### PR DESCRIPTION
**Pull request by richardpen**
### Summary

Fix #1151
Previously all the plugins were invoked in one call in agent `AddNetworkList` with the same runtime configuration, this PR separates them into different calls with `AddNetwork` so that each plugin can have different runtime config.

### Implementation details

Replace the `AddNetworkList` with `AddNetwork` and call the plugin in the order of `ecs-eni` and then `ecs-bridge`.

### Testing

- [x] Builds on Linux (`make release`)
- [x] Builds on Windows (`go build -out amazon-ecs-agent.exe ./agent`)
- [x] Unit tests on Linux (`make test`) pass
- [x] Unit tests on Windows (`go test -timeout=25s ./agent/...`) pass
- [x] Integration tests on Linux (`make run-integ-tests`) pass
- [x] Integration tests on Windows (`.\scripts\run-integ-tests.ps1`) pass
- [x] Functional tests on Linux (`make run-functional-tests`) pass
- [x] Functional tests on Windows (`.\scripts\run-functional-tests.ps1`) pass
- [x] Checked out the latest `amazon-ecs-cni-plugin`(including aws/amazon-ecs-cni-plugins#59) and run a task with `awsvpc` mode, the container namespace was set correctly.
New tests cover the changes: 

### Licensing

This contribution is under the terms of the Apache 2.0 License: 
yes

### CODE REVIEW

File: agent/ecscni/plugin.go [hunk 1]
```diff
@@ -70,37 +70,36 @@ func NewClient(cfg *Config) CNIClient {
...
 
-	seelog.Debugf("Set up container namespace done: %s", result.String())
+	result, err = client.bridgeAdd(rtf, cfg)
+	if err != nil {
+		return nil, errors.Wrap(err, "cni setup: invoke bridge plugin failed")
+	}
+	seelog.Debugf("ecscni: Set up container namespace done: %s", result.String())
 	if _, err = result.GetAsVersion(currentCNISpec); err != nil {
-		seelog.Warnf("Unable to convert result to %s; error: %v; result is of version: %s",
+		seelog.Warnf("ecscni: Unable to convert result to %s; error: %v; result is of version: %s",
```
Comment (adnxn) [inline agent/ecscni/plugin.go]: nit: `Unable to convert result to spec version %s`?

File: agent/ecscni/plugin.go [hunk 2]
```diff
@@ -70,37 +70,36 @@ func NewClient(cfg *Config) CNIClient {
...
 
-	seelog.Debugf("Set up container namespace done: %s", result.String())
+	result, err = client.bridgeAdd(runtimeConfig, cfg)
+	if err != nil {
+		return nil, errors.Wrap(err, "cni setup: invoke bridge plugin failed")
+	}
+	seelog.Debugf("ecscni: Set up container namespace done: %s", result.String())
 	if _, err = result.GetAsVersion(currentCNISpec); err != nil {
-		seelog.Warnf("Unable to convert result to %s; error: %v; result is of version: %s",
+		seelog.Warnf("ecscni: Unable to convert result to spec version %s; error: %v; result is of version: %s",
```
Comment (aaithal) [inline agent/ecscni/plugin.go]: please make sure that log statements start with upper case letters

File: agent/ecscni/types.go [hunk 3]
```diff
@@ -24,7 +24,9 @@ const (
 	// CNI plugin
 	capabilitiesCommand = "--capabilities"
 	// defaultEthName is the name of veth pair name in the container namespace
```
Comment (aaithal) [inline agent/ecscni/types.go]: nit: defaultVethName

File: agent/ecscni/plugin.go [hunk 4]
```diff
@@ -110,92 +109,115 @@ func (client *cniClient) SetupNS(cfg *Config) (*current.Result, error) {
...
-	return client.libcni.DelNetworkList(networkConfigList, cns)
+	return client.libcni.DelNetwork(ipamConfig, runtimeConfig)
 }
 
-// createNetworkConfig creates the configuration for invoking plugins
-func (client *cniClient) createNetworkConfig(cfg *Config, bridgeConfigFunc func(*Config) (*libcni.NetworkConfig, error)) (*libcni.NetworkConfigList, error) {
-	bridgeConfig, err := bridgeConfigFunc(cfg)
+// bridgeAdd invokes the bridge and ipam plugin to set up the bridge,
+// also it sets the veth pair and route in container namespace
+func (client *cniClient) bridgeAdd(runtimeConfig libcni.RuntimeConf, cfg *Config) (cnitypes.Result, error) {
```
Comment (aaithal) [inline agent/ecscni/plugin.go]: Can you combine `bridgeAdd` and `eniAdd` into a single method? I think you can pass `createBridgeNetworkConfigWithIPAM` and `createENINetworkConfig` as parameters to this method

File: agent/ecscni/plugin.go [hunk 5]
```diff
@@ -110,92 +109,115 @@ func (client *cniClient) SetupNS(cfg *Config) (*current.Result, error) {
...
-	pluginConfigs := []*libcni.NetworkConfig{eniConfig, bridgeConfig}
-	networkConfigList := &libcni.NetworkConfigList{
-		CNIVersion: client.cniVersion,
-		Plugins:    pluginConfigs,
+	runtimeConfig.IfName = deviceName
+	return client.libcni.AddNetwork(networkConfig, &runtimeConfig)
+}
+
+// bridgeDel invokes the bridge plugin to clean up the container namespace
+func (client *cniClient) bridgeDel(runtimeConfig libcni.RuntimeConf, cfg *Config) error {
```
Comment (aaithal) [inline agent/ecscni/plugin.go]: Can you combine `bridgeDel` and `eniDel` into a single method similar to my other comment?

File: agent/ecscni/plugin.go [hunk 6]
```diff
@@ -70,37 +70,38 @@ func NewClient(cfg *Config) CNIClient {
...
 	defer os.Unsetenv("ECS_CNI_LOGLEVEL")
-	result, err := client.libcni.AddNetworkList(networkConfigList, cns)
+
+	// Invoke eni plugin ADD command
+	result, err := client.add(runtimeConfig, cfg, client.createENINetworkConfig)
 	if err != nil {
-		return nil, err
+		return nil, errors.Wrap(err, "cni setup: invoke eni plugin failed")
 	}
+	seelog.Debugf("Ecscni: Set up eni done: %s", result.String())
```
Comment (aaithal) [inline agent/ecscni/plugin.go]: nit: `Ecs` -> `ECS` everywhere.

File: agent/ecscni/plugin.go [hunk 7]
```diff
@@ -241,23 +242,25 @@ func (client *cniClient) createENINetworkConfig(cfg *Config) (*libcni.NetworkCon
 		IPV6Address:          cfg.ENIIPV6Address,
 		BlockInstanceMetdata: cfg.BlockInstanceMetdata,
 	}
-
-	return client.constructNetworkConfig(eniConf, ECSENIPluginName)
+	networkConfig, err := client.constructNetworkConfig(eniConf, ECSENIPluginName)
+	return defaultENIName, networkConfig, err
```
Comment (aaithal) [inline agent/ecscni/plugin.go]: can you add `if err != nil` check here and make sure that you're wrapping the error as well?

File: agent/ecscni/plugin.go [hunk 8]
```diff
@@ -110,92 +111,92 @@ func (client *cniClient) SetupNS(cfg *Config) (*current.Result, error) {
...
+	networkConfig, err := client.constructNetworkConfig(bridgeConfig, ECSBridgePluginName)
+	return defaultVethName, networkConfig, err
 }
 
 // createBridgeNetworkConfigWithoutIPAM creates the config of the bridge for removal
-func (client *cniClient) createBridgeNetworkConfigWithoutIPAM(cfg *Config) (*libcni.NetworkConfig, error) {
-	return client.constructNetworkConfig(client.createBridgeConfig(cfg), ECSBridgePluginName)
+func (client *cniClient) createBridgeNetworkConfigWithoutIPAM(cfg *Config) (string, *libcni.NetworkConfig, error) {
+	networkConfig, err := client.constructNetworkConfig(client.createBridgeConfig(cfg), ECSBridgePluginName)
+	return defaultVethName, networkConfig, err
```
Comment (aaithal) [inline agent/ecscni/plugin.go]: can you add `if err != nil` check here and make sure that you're wrapping the error as well?

**Comment by richardpen** [comment]
@sharanyad Yeah, I have linked them in the description.

**Pull request merged**